operator/internal/manifests/prometheus_rule.go

```diff
@@ -0,0 +1,62 @@
+package manifests
+
+import (
+	"bytes"
+
+	"github.com/ViaQ/logerr/kverrors"
+	"github.com/grafana/loki/operator/internal/manifests/internal/alerts"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sYAML "k8s.io/apimachinery/pkg/util/yaml"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// BuildPrometheusRule returns a list of k8s objects for Loki PrometheusRule
+func BuildPrometheusRule(opts Options) ([]client.Object, error) {
+	prometheusRule, err := NewPrometheusRule(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return []client.Object{
+		prometheusRule,
+	}, nil
+}
+
+// NewPrometheusRule creates a prometheus rule
+func NewPrometheusRule(opts Options) (*monitoringv1.PrometheusRule, error) {
+	alertsBytes, err := alerts.Build()
+	if err != nil {
+		return nil, err
+	}
+
+	var alertsSpec *monitoringv1.PrometheusRuleSpec
+	alertsSpec, err = ruleSpec(alertsBytes)
```

Comment:
I tried to follow the same pattern as in `gateway` and `config` where the subpackage `Build()` method returns `[]byte` and all the other processing is done in package `manifests`. Do you still prefer to move this logic to the subpackage?

Reply:
I added a commit to address the second point of parsing the YAML once.

Reply:
> I tried to follow the same pattern as in `gateway` and `config` where the subpackage `Build()` method returns `[]byte` and all the other processing is done in package `manifests`. Do you still prefer to move this logic to the subpackage?

The main difference is that the output of both of those packages is not touched again but just added to a ConfigMap. In that case `[]byte` seems like a sensible choice, because that is what is needed in the ConfigMap.

In the case of this PR the output is immediately handed off to a YAML parser instead. I think at least doing the YAML parsing can still be done in what's currently `alerts` package, moving most of the PrometheusRule-related code that is now in manifests to that package.

Reply:
OK, Done.

Reply:
That's the right direction, but I think we can go a step further. Currently the "caching logic" is in the manifests package as well, but I would think that this is an implementation detail of the current "alerts template" and thus also belongs into the `alerts` package.

The caching logic itself could also be simplified by having the read/parse happen during `init()` of the package as long as there are no moving parts in there. The downside would be that the parse will happen during startup of the controller and even if the user is not using the alerts at all. On the other side we would guarantee a failure if the parse is not possible and seeing that the current template is completely static and out of the control of the user it seems like a good idea to fail early.